File: python/src/anni/requests/requests.py
from typing import List
from pydantic import TypeAdapter
from anni.requests.responsetypes import (
    CheckElfLogicRequestPerson,
    ScoreElfLogicRequest,
    ScoreElfLogicRequestPerson,
    ScoreElfLogicResponse,
)
from anni.types.datatypes import NaughtyOrNiceClassificationEnum, Person, ScoredPerson
from anni.types.settings import Settings
import requests

class RequestHandler:
    def __init__(self, settings: Settings | None = None, session: requests.Session | None = None):
        self.settings = settings or Settings()
        self.session = session or requests.Session()
        self.check_elf_logic_adaptor = TypeAdapter(List[CheckElfLogicRequestPerson])
        self.elf_check_response_adaptor = TypeAdapter(NaughtyOrNiceClassificationEnum)
        self.elf_logic_request_adaptor = TypeAdapter(List[ScoreElfLogicRequestPerson])

    def _get_team_name(self):
        if not self.settings.team_name:
            raise ValueError("Team name is not set in settings.")
        return self.settings.team_name

    def _get_team_secret(self):
        if not self.settings.team_secret:
            raise ValueError("Unique key is not set in settings.")
        return self.settings.team_secret

    def _get_auth_headers(self):
        return {
            "X-Team-Name": self._get_team_name(),
            "X-Unique-Key": self._get_team_secret(),
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

    def _url(self, endpoint: str) -> str:
        return f"{self.settings.base_url.rstrip('/')}/{endpoint.lstrip('/')}"

    def _post(self, endpoint: str, json_data: dict):
        response = self.session.post(
            self._url(endpoint),
            json=json_data,
            headers=self._get_auth_headers(),
            timeout=10,
        )
        print(response.request.body)
        response.raise_for_status()
        return response.json()
    
    def _getSingle(self, endpoint: str, json_data: dict):
        response = self.session.get(
            self._url(endpoint),
            json=json_data,
            headers=self._get_auth_headers(),
            timeout=10,
        )
        response.raise_for_status()
        return response.json()

    # Public API
    def check_elf_logic(self, person: Person):
        endpoint = "naughtyornice"
        request_data = CheckElfLogicRequestPerson(**person.model_dump()).model_dump()
        parsed_response = self._getSingle(endpoint, request_data)
        return self.elf_check_response_adaptor.validate_python(parsed_response)

    def submit_list(self, people: List[ScoredPerson]):
        endpoint = "score"
        request_data = ScoreElfLogicRequest(
            naughtyNiceCandidateRequests=[ScoreElfLogicRequestPerson(id=p.id, naughtyOrNice=p.naughty_or_nice) for p in people]
        ).model_dump()

        parsed_response = self._post(endpoint, request_data)
        return ScoreElfLogicResponse.model_validate(parsed_response)

File: javascript/src/anni/requests/requests.js
import axios from 'axios';
import { Settings } from '../types/settings.js';
import { 
  ScoreElfLogicRequestSchema, 
  ScoreElfLogicResponseSchema 
} from './responsetypes.js';
import { NaughtyOrNiceClassificationEnum } from '../types/datatypes.js';
import { z } from 'zod';

export class RequestHandler {
  constructor(settings = null, session = null) {
    this.settings = settings || new Settings();
    this.session = session || axios.create();
  }

  _getTeamName() {
    if (!this.settings.team_name) {
      throw new Error("Team name is not set in settings.");
    }
    return this.settings.team_name;
  }

  _getTeamSecret() {
    if (!this.settings.team_secret) {
      throw new Error("Unique key is not set in settings.");
    }
    return this.settings.team_secret;
  }

  _getAuthHeaders() {
    return {
      "X-Team-Name": this._getTeamName(),
      "X-Unique-Key": this._getTeamSecret(),
      "Content-Type": "application/json",
      "Accept": "application/json"
    };
  }

  _url(endpoint) {
    const baseUrl = this.settings.base_url.replace(/\/$/, '');
    const cleanEndpoint = endpoint.replace(/^\//, '');
    return `${baseUrl}/${cleanEndpoint}`;
  }

  async _post(endpoint, jsonData) {
    try {
      const response = await this.session.post(
        this._url(endpoint),
        jsonData,
        {
          headers: this._getAuthHeaders(),
          timeout: 10000
        }
      );
      return response.data;
    } catch (error) {
      if (error.response) {
        throw new Error(`HTTP ${error.response.status}: ${error.response.statusText}`);
      }
      throw error;
    }
  }

  async _getSingle(endpoint, jsonData) {
    try {
      const response = await this.session.request({
        method: 'GET',
        url: this._url(endpoint),
        data: jsonData,
        headers: this._getAuthHeaders(),
        timeout: 10000
      });
      return response.data;
    } catch (error) {
      if (error.response) {
        throw new Error(`HTTP ${error.response.status}: ${error.response.statusText}`);
      }
      throw error;
    }
  }

  // Public API
  async checkElfLogic(person) {
    const endpoint = "naughtyornice";
    const requestData = { ...person };
    const parsedResponse = await this._getSingle(endpoint, requestData);
    
    // Validate response is a valid NaughtyOrNiceClassificationEnum value
    const enumSchema = z.enum([
      NaughtyOrNiceClassificationEnum.NONE,
      NaughtyOrNiceClassificationEnum.NAUGHTY,
      NaughtyOrNiceClassificationEnum.NICE
    ]);
    
    return enumSchema.parse(parsedResponse);
  }

  async submitList(people) {
    const endpoint = "score";
    const requestData = {
      naughtyNiceCandidateRequests: people.map(p => ({
        id: p.id,
        naughtyOrNice: p.naughty_or_nice
      }))
    };
    
    // Validate request data
    ScoreElfLogicRequestSchema.parse(requestData);
    
    const response = await this._post(endpoint, requestData);
    const parsedResponse = ScoreElfLogicResponseSchema.parse(response);
    
    return parsedResponse
  }
}


